DOC: 1.10.0 release notes (#17534)
scipy/scipy: doc/release/1.10.0-notes.rst

Highlights of this release
**************************

- a new dedicated datasets submodule (`scipy.datasets`) has been added, and is
  now preferred over usage of `scipy.misc` for dataset retrieval

************
New features
************

`scipy.datasets` introduction
=============================
- A new dedicated datasets submodule moving all the dataset methods from
  `scipy.misc` has been added.
- The new submodule includes: `scipy.datasets.face`, `scipy.datasets.ascent`,
  `scipy.datasets.electrocardiogram`
- The submodule is based on ``Pooch`` (an optional dependency for SciPy), a
  Python package to simplify fetching data files. This move will facilitate
  SciPy to cut the sdist/wheel in size by decoupling the data files and moving
  them out of the SciPy repository, hosting them externally and downloading
  them when requested. After downloading the datasets once, the files are
  cached to avoid network dependence and repeated usage.

`scipy.integrate` improvements
==============================

> - Added `scipy.integrate.qmc_quad`, which uses performs quadrature using

Author: `uses performs` typo copied over from wiki I think

  Quasi-Monte Carlo points.
- Added parameter ``complex_func`` to `scipy.integrate.quad`, which can be set
  ``True`` to integrate a complex integrand.

`scipy.cluster` improvements
============================

`scipy.interpolate` improvements
================================
- `scipy.interpolate.interpn` now supports tensor-product interpolation methods
  (``slinear``, ``cubic``, ``quintic`` and ``pchip``)
- Tensor-product interpolation methods (``slinear``, ``cubic``, ``quintic`` and
  ``pchip``) in `scipy.interpolate.interpn` and
  `scipy.interpolate.RegularGridInterpolator` now allow values with trailing
  dimensions.
- `scipy.interpolate.RegularGridInterpolator` has a new fast path for
  ``method="linear"`` with 2D data, and ``RegularGridInterpolator`` is now
  easier to subclass
- `scipy.interpolate.interp1d` now can take a single value for non-spline
  methods.
- A new ``extrapolate`` argument is available to ``BSpline.design_matrix``,
  allowing extrapolation based on the first and last intervals.
- A new function ``make_smoothing_spline`` has been added. It is an
  implementation of the generalized cross-validation spline smoothing

>   algorithm.

Reviewer: ```suggestion
  algorithm. The ``lam=None`` (default) mode of this function is a clean-room reimplementation of the classic ``gcvspl.f`` Fortran algorithm for constructing GCV splines.
  
- A new ``method="pchip"`` mode was aded to   `scipy.interpolate.RegularGridInterpolator`. This mode constructs an interpolator using tensor products of C1-continuous monotone splines (essentially, a `scipy.interpolate.PchipInterpolator` instance per dimension).
```

  them when requested. After downloading the datasets once, the files are
  cached to avoid network dependence and repeated usage.

`scipy.integrate` improvements
==============================
- Added `scipy.integrate.qmc_quad`, which performs quadrature using Quasi-Monte
  Carlo points.
- Added parameter ``complex_func`` to `scipy.integrate.quad`, which can be set
  ``True`` to integrate a complex integrand.

`scipy.cluster` improvements
============================

`scipy.interpolate` improvements
================================
- `scipy.interpolate.interpn` now supports tensor-product interpolation methods
  (``slinear``, ``cubic``, ``quintic`` and ``pchip``)
- Tensor-product interpolation methods (``slinear``, ``cubic``, ``quintic`` and
  ``pchip``) in `scipy.interpolate.interpn` and
  `scipy.interpolate.RegularGridInterpolator` now allow values with trailing
  dimensions.
- `scipy.interpolate.RegularGridInterpolator` has a new fast path for
  ``method="linear"`` with 2D data, and ``RegularGridInterpolator`` is now
  easier to subclass
- `scipy.interpolate.interp1d` now can take a single value for non-spline
  methods.

> - A new ``extrapolate`` argument is available to ``BSpline.design_matrix``,

Reviewer 2: I'm not sure if this will render
```suggestion
- A new ``extrapolate`` argument is available to `scipy.interpolate.BSpline.design_matrix`,
```
or if it needs to be 
```suggestion
- A new ``extrapolate`` argument is available to the ``design_matrix`` method of `scipy.interpolate.BSpline`,
```

  allowing extrapolation based on the first and last intervals.

> - A new function ``make_smoothing_spline`` has been added. It is an

Reviewer 2: There are probably dozens of places where double-backticks could be changed to single backticks with fully-qualified names so that the links render. ~~LMK if you want me to go through all of these.~~ I'll go ahead and hunt them down, because I think it's helpful to have the link for those who want to learn more.
```suggestion
- A new function `scipy.interpolate.make_smoothing_spline` has been added. It is an
```

  implementation of the generalized cross-validation spline smoothing
  algorithm. The ``lam=None`` (default) mode of this function is a clean-room
  reimplementation of the classic ``gcvspl.f`` Fortran algorithm for
  constructing GCV splines.
- A new ``method="pchip"`` mode was aded to
  `scipy.interpolate.RegularGridInterpolator`. This mode constructs an
  interpolator using tensor products of C1-continuous monotone splines
  (essentially, a `scipy.interpolate.PchipInterpolator` instance per
  dimension).

`scipy.sparse.linalg` improvements
==================================
- The spectral 2-norm is now available in `scipy.sparse.linalg.norm`.
- The performance of `scipy.sparse.linalg.norm` for the default case (Frobenius
  norm) has been improved.
- LAPACK wrappers were added for ``trexc`` and ``trsen``.

`scipy.linalg` improvements
===========================

> - ``lu_factor`` now accepts rectangular arrays instead of being restricted

Reviewer 2: ```suggestion
- `scipy.linalg.lu_factor` now accepts rectangular arrays instead of being restricted
```

  to square arrays.

`scipy.ndimage` improvements
============================

> - The new ``value_indices()`` function provides a time-efficient method to

Reviewer 2: ```suggestion
- The new `scipy.ndimage.value_indices` function provides a time-efficient method to
```

  search for the locations of individual values with an array of image data.

> - A new ``radius`` argument is supported by ``gaussian_filter1d`` and
>   ``gaussian_filter`` for adjusting the kernel size of the filter.

Reviewer 2: ```suggestion
- A new ``radius`` argument is supported by `scipy.ndimage.gaussian_filter1d` and
  `scipy.ndimage.gaussian_filter` for adjusting the kernel size of the filter.
```

`scipy.optimize` improvements
=============================
- `scipy.optimize.brute` now coerces non-iterable/single-value ``args`` into a
  tuple.
- `scipy.optimize.least_squares` and `scipy.optimize.curve_fit` now accept
  `scipy.optimize.Bounds` for bounds constraints.
- Added a tutorial for `scipy.optimize.milp`.
- Improved the pretty-printing of `scipy.optimize.OptimizeResult` objects.

> - some additional options (``parallel``, ``threads``, ``mip_rel_gap``) can now
>   be properly passed to the HiGHS solver

Reviewer 2: ```suggestion
- Additional options (``parallel``, ``threads``, ``mip_rel_gap``) can now
  be passed to `scipy.optimize.linprog` with ``method='highs'``.
```

`scipy.signal` improvements
===========================
- The new window function `scipy.signal.windows.lanczos` was added to compute a
  Lanczos window, also known as a sinc window.

`scipy.sparse` improvements
===========================

`scipy.sparse.csgraph` improvements
===================================
- the performance of `scipy.sparse.csgraph.dijkstra` has been improved, and
  star graphs in particular see a marked performance improvement

`scipy.spatial` improvements
============================

`scipy.special` improvements
============================
- The new function `scipy.special.powm1`, a ufunc with signature
  ``powm1(x, y)``, computes ``x**y - 1``. The function avoids the loss of
  precision that can result when ``y`` is close to 0 or when ``x`` is close to
  1.

> - ``erfinv`` is now more accurate as it leverages the Boost equivalent under

Reviewer 2: ```suggestion
- `scipy.special.erfinv` is now more accurate as it leverages the Boost equivalent under
```

  the hood.

`scipy.stats` improvements
==========================
- Added `scipy.stats.goodness_of_fit`, a generalized goodness-of-fit test for
  use with any univariate distribution, any combination of known and unknown
  parameters, and several choices of test statistic (Kolmogorov-Smirnov,
  Cramer-von Mises, and Anderson-Darling).
- Improved `scipy.stats.bootstrap`: Default method ``'BCa'`` now supports
  multi-sample statistics. Also, the bootstrap distribution is returned in the
  result object, and the result object can be passed into the function to add
  additional resamples or change the confidence interval level and type.
- Added a ``confidence_interval`` method to the result object returned by
  `scipy.stats.ttest_1samp` and `scipy.stats.ttest_rel``.
- Added maximum spacing estimation to `scipy.stats.fit`.

> - Added `scipy.stats.oddsratio` to compute both the conditional and

Reviewer 2: ```suggestion
- Added `scipy.stats.contingency.odds_ratio` to compute both the conditional and
```

  unconditional odds ratio for 2x2 contingency tables and corresponding
  confidence intervals.
- Added the Poisson means test ("E-test") as `scipy.stats.poisson_means_test`.
- Added `scipy.stats.expectile`, which generalizes the expected value in the
  same way as quantiles are a generalization of the median.
- Added new statistical distributions.

>   * Added `scipy.stats.uniform_direction`, a multivariate distribution to
>   sample uniformly from the surface of a hypersphere.
>   * Added `scipy.stats.random_table`, a multivariate distribution to sample
>   uniformly from m x n contingency tables with provided marginals.
>   * Added `scipy.stats.truncpareto`, the truncated Pareto distribution.
> - Improved the ``fit`` method of several distributions.
>   * `scipy.stats.skewnorm` and `scipy.stats.weibull_min` now use an analytical
>   solution when ``method='mm'``, which also serves a starting guess to improve
>   the performance of ``method='mle'``.
>   * `scipy.stats.gumbel_r` and `scipy.stats.gumbel_l`: analytical maximum
>   likelihood estimates have been extended to the cases in which location or
>   scale are fixed by the user.
>   * Analytical maximum likelihood estimates have been added for
>   `scipy.stats.powerlaw`.
> - Improved random variate sampling of several distributions.
>   * Drawing multiple samples from `scipy.stats.matrix_normal`,
>   `scipy.stats.ortho_group`, `scipy.stats.special_ortho_group`, and
>   `scipy.stats.unitary_group` is faster.
>   * The ``rvs`` method of `scipy.stats.vonmises` now wraps to the interval
>   ``[-np.pi, np.pi]``
>   * Improved the reliability of `scipy.stats.loggamma` ``rvs`` method for small
>   values of the shape parameter.

Reviewer 2: Oops. Hopefully this nests the bullets as desired.
```suggestion
  - Added `scipy.stats.uniform_direction`, a multivariate distribution to
  sample uniformly from the surface of a hypersphere.
  - Added `scipy.stats.random_table`, a multivariate distribution to sample
  uniformly from m x n contingency tables with provided marginals.
  - Added `scipy.stats.truncpareto`, the truncated Pareto distribution.
- Improved the ``fit`` method of several distributions.
  - `scipy.stats.skewnorm` and `scipy.stats.weibull_min` now use an analytical
  solution when ``method='mm'``, which also serves a starting guess to improve
  the performance of ``method='mle'``.
  - `scipy.stats.gumbel_r` and `scipy.stats.gumbel_l`: analytical maximum
  likelihood estimates have been extended to the cases in which location or
  scale are fixed by the user.
  - Analytical maximum likelihood estimates have been added for
  `scipy.stats.powerlaw`.
- Improved random variate sampling of several distributions.
  - Drawing multiple samples from `scipy.stats.matrix_normal`,
  `scipy.stats.ortho_group`, `scipy.stats.special_ortho_group`, and
  `scipy.stats.unitary_group` is faster.
  - The ``rvs`` method of `scipy.stats.vonmises` now wraps to the interval
  ``[-np.pi, np.pi]``
  - Improved the reliability of `scipy.stats.loggamma` ``rvs`` method for small
  values of the shape parameter.
```

- Improved the speed and/or accuracy of functions of several statistical
  distributions.

>   * Added `scipy.stats.Covariance` for better speed, accuracy, and user control
>   in multivariate normal calculations.
>   * `scipy.stats.skewnorm` methods ``cdf``, ``sf``, ``ppf``, and ``isf``
>   methods now use the implementations from Boost, improving speed while
>   maintaining accuracy. The calculation of higher-order moments is also faster
>   and more accurate.
>   * `scipy.stats.invgauss` methods ``ppf`` and ``isf`` methods now use the
>   implementations from Boost, improving speed and accuracy.
>   * `scipy.stats.invweibull` methods ``sf`` and ``isf`` are more accurate for
>   small probability masses.
>   * `scipy.stats.nct` and `scipy.stats.ncx2` now rely on the implementations
>   from Boost, improving speed and accuracy.
>   * Implemented the ``logpdf`` method of `scipy.stats.vonmises` for reliability
>   in extreme tails.
>   * Implemented the ``isf`` method of `scipy.stats.levy` for speed and
>   accuracy.
>   * Improved the robustness of `scipy.stats.studentized_range` for large ``df``
>   by adding an infinite degree-of-freedom approximation.
>   * Added a parameter ``lower_limit`` to ``stats.multivariate_normal``,
>   allowing the user to change the integration limit from -inf to a desired
>   value.
>   * Improved the robustness of ``entropy`` of `scipy.stats.vonmises` for large
>   concentration values
> - Enhanced `scipy.stats.gaussian_kde`.
>   * Added `scipy.stats.gaussian_kde.marginal`, which returns the desired
>   marginal distribution of the original kernel density estimate distribution.
>   * The ``cdf`` method of `scipy.stats.gaussian_kde` now accepts a
>   ``lower_limit`` parameter for integrating the PDF over a rectangular region.
>   * Moved calculations for `scipy.stats.gaussian_kde.logpdf` to Cython,
>   improving speed.
>   * The global interpreter lock is released by the ``pdf`` method of
>   `scipy.stats.gaussian_kde` for improved multithreading performance.
>   * Replaced explicit matrix inversion with Cholesky decomposition for speed
>   and accuracy.

Reviewer 2: ```suggestion
  - Added `scipy.stats.Covariance` for better speed, accuracy, and user control
  in multivariate normal calculations.
  - `scipy.stats.skewnorm` methods ``cdf``, ``sf``, ``ppf``, and ``isf``
  methods now use the implementations from Boost, improving speed while
  maintaining accuracy. The calculation of higher-order moments is also faster
  and more accurate.
  - `scipy.stats.invgauss` methods ``ppf`` and ``isf`` methods now use the
  implementations from Boost, improving speed and accuracy.
  - `scipy.stats.invweibull` methods ``sf`` and ``isf`` are more accurate for
  small probability masses.
  - `scipy.stats.nct` and `scipy.stats.ncx2` now rely on the implementations
  from Boost, improving speed and accuracy.
  - Implemented the ``logpdf`` method of `scipy.stats.vonmises` for reliability
  in extreme tails.
  - Implemented the ``isf`` method of `scipy.stats.levy` for speed and
  accuracy.
  - Improved the robustness of `scipy.stats.studentized_range` for large ``df``
  by adding an infinite degree-of-freedom approximation.
  - Added a parameter ``lower_limit`` to `scipy.stats.multivariate_normal`,
  allowing the user to change the integration limit from -inf to a desired
  value.
  - Improved the robustness of ``entropy`` of `scipy.stats.vonmises` for large
  concentration values
- Enhanced `scipy.stats.gaussian_kde`.
  - Added `scipy.stats.gaussian_kde.marginal`, which returns the desired
  marginal distribution of the original kernel density estimate distribution.
  - The ``cdf`` method of `scipy.stats.gaussian_kde` now accepts a
  ``lower_limit`` parameter for integrating the PDF over a rectangular region.
  - Moved calculations for `scipy.stats.gaussian_kde.logpdf` to Cython,
  improving speed.
  - The global interpreter lock is released by the ``pdf`` method of
  `scipy.stats.gaussian_kde` for improved multithreading performance.
  - Replaced explicit matrix inversion with Cholesky decomposition for speed
  and accuracy.
```

- Statistical test functions that previously returned plain tuples now return
  bunches, allowing attributes to be accessed by name. The `scipy.stats`
  functions ``combine_pvalues``, ``fisher_exact``, ``chi2_contingency``,
  ``median_test`` and ``mood`` have been updated.
- The result objects returned by statistical test functions now consistently
  use the attribute names ``statistic`` and ``pvalue``. Old attribute names are
  allowed for backward compatibility. The `scipy.stats` functions
  ``multiscale_graphcorr``, ``anderson_ksamp``, ``binomtest``, ``crosstab``,
  ``pointbiserialr``, ``spearmanr``, ``kendalltau`` and ``weightedtau`` have
  been updated.
- `scipy.stats.anderson` now returns the parameters of the fitted distribution
  in a `scipy.stats._result_classes.FitResult` object.
- Kolmogorov-Smirnov tests (e.g. `scipy.stats.kstest`) return more information:
  the location (argmax) at which the statistic is calculated, and the variant
  of the statistic used.
- Improved the performance of `scipy.stats.cramervonmises_2samp` and
  `scipy.stats.ks_2samp` with ``method='exact'``.
- Improved the performance of `scipy.stats.siegelslopes`.

> - Improved the performance of `stats.hdquantile_sd`
> - Added the ``scramble`` optional argument to `stats.qmc.LatinHypercube`. It

Reviewer 2: ```suggestion
- Improved the performance of `scipy.stats.hdquantile_sd`
- Added the ``scramble`` optional argument to `scipy.stats.qmc.LatinHypercube`. It
```

  replaces ``centered`` which is now deprecated.
- Added a parameter ``optimization`` to all `scipy.stats.qmc.QMCEngine`
  subclasses to improve characteristics of the quasi-random variates
- Added `scipy.stats.directional_stats` to compute sample statistics of
  n-dimensional directional data.
- The ``plot`` method of `scipy.stats._result_classes.FitResult` now accepts a
  ``plot_type`` parameter; the options are ``'hist'`` (histogram, default),
  ``'qq'`` (Q-Q plot), ``'pp'`` (P-P plot), and ``'cdf'`` (empirical CDF plot).
- Added tie correction to `scipy.stats.mood`.
- Added tutorials for resampling methods in `scipy.stats`.
- `scipy.stats.bootstrap`, `scipy.stats.permutation_test`, and
  `scipy.stats.monte_carlo_test` now automatically detect whether the provided
  ``statistic`` is vectorized, so passing the ``vectorized`` argument
  explicitly is no longer required to take advantage of vectorized statistics.
- Improved speed of `scipy.stats.permutation_test` for permutation types
  ``'samples'`` and ``'pairings'``.

> - Improved the performance of `scipy.stats.binned_statistics_dd` for several

Reviewer 2: ```suggestion
- Improved the performance of `scipy.stats.binned_statistic_dd` for several
```

  NumPy statistics, and binned statistics methods now support complex data.

> - Added ``axis``, ``nan_policy``, and masked array support to
>   `stats.jarque_bera`
> - Added the ``nan_policy`` optional argument to `stats.rankdata`.

Reviewer 2: ```suggestion
- Added ``axis``, ``nan_policy``, and masked array support to
  `scipy.stats.jarque_bera`
- Added the ``nan_policy`` optional argument to `scipy.stats.rankdata`.
```

*******************
Deprecated features
*******************
- `scipy.misc` module and all the methods in ``misc`` are deprecated in v1.10
  and will be completely removed in SciPy v2.0.0. Users are suggested to
  utilize the `scipy.datasets` module instead for the dataset methods.
- `scipy.stats.qmc.LatinHypercube` parameter ``centered`` has been deprecated.
  It is replaced by the ``scramble`` argument for more consistency with other

>   QMC engines.

Reviewer: ```suggestion
  QMC engines.
  
  - `scipy.interpolate.interp2d` class has been deprecated.  The docstring of the deprecated routine lists recommended replacements.
```

*******************
Deprecated features
*******************
- `scipy.misc` module and all the methods in ``misc`` are deprecated in v1.10
  and will be completely removed in SciPy v2.0.0. Users are suggested to
  utilize the `scipy.datasets` module instead for the dataset methods.
- `scipy.stats.qmc.LatinHypercube` parameter ``centered`` has been deprecated.
  It is replaced by the ``scramble`` argument for more consistency with other
  QMC engines.
- `scipy.interpolate.interp2d` class has been deprecated.  The docstring of the
  deprecated routine lists recommended replacements.

********************
Expired Deprecations
********************
- There is an ongoing effort to follow through on long-standing deprecations.
  The following previously deprecated features are affected:

>   * Removed ``cond`` & ``rcond`` kwargs in ``linalg.pinv``
>   * Removed wrappers ``scipy.linalg.blas.{clapack, flapack}``

Reviewer 3: Sorry that the deprecation release notes were not up to date...

```suggestion
  * Removed ``cond`` & ``rcond`` kwargs in ``linalg.pinv``
  * Removed wrappers ``scipy.linalg.blas.{clapack, flapack}``
  * Removed ``scipy.stats.NumericalInverseHermite`` and removed ``tol`` & ``max_intervals`` kwargs from ``scipy.stats.sampling.NumericalInverseHermite``
  * Removed ``local_search_options`` kwarg frrom ``scipy.optimize.dual_annealing``.
```

* GavinZhang (2) +
* David Gilbertson (1) +
* Ralf Gommers (251)
* Marco Gorelli (2) +
* Matt Haberland (373)
* Andrew Hawryluk (2) +
* Loïc Houpert (2) +
* Shamus Husheer (1) +
* ideasrule (1) +
* imoiwm (1) +
* Lakshaya Inani (1) +
* Joseph T. Iosue (1)
* iwbc-mzk (1) +
* Nathan Jacobi (3) +
* Julien Jerphanion (5)
* He Jia (1)
* jmkuebler (1) +
* Johannes Müller (1) +
* Vedant Jolly (1) +
* Juan Luis Cano Rodríguez (2)
* Justin (1) +
* jvavrek (1) +
* jyuv (2)
* Kai Mühlbauer (1) +
* Nikita Karetnikov (2) +
* Reinert Huseby Karlsen (1) +
* kaspar (2) +
* Toshiki Kataoka (1)
* Robert Kern (3)
* Joshua Klein (1) +
* Andrew Knyazev (5)
* korneix (1) +
* Jozsef Kutas (16) +
* Eric Larson (4)
* Lechnio (1) +
* Antony Lee (2)
* Aditya Limaye (1) +
* Xingyu Liu (2)
* Christian Lorentzen (4)
* Loïc Estève (2)
* Thibaut Lunet (2) +
* Peter Lysakovski (1)
* marianasalamoni (2) +
* mariprudencio (1) +
* Paige Martin (1) +
* matthewborish (3) +
* Damon McDougall (1)
* Nicholas McKibben (22)
* McLP (1) +
* mdmahendri (1) +
* Melissa Weber Mendonça (9)
* Jarrod Millman (1)
* Naoto Mizuno (2)
* Shashaank N (1)
* Pablo S Naharro (1) +
* nboudrie (1) +
* Andrew Nelson (50)
* Nico Schlömer (1)

> * Nikita Karetnikov (ニキータ カレートニコフ) (1) +

Reviewer: duplicated: this could be a `.mailmap` entry

*************
Other changes
*************
- `scipy.stats.bootstrap`, `scipy.stats.permutation_test`, and
  `scipy.stats.monte_carlo_test` now automatically detect whether the provided
  ``statistic`` is vectorized by looking for an ``axis`` parameter in the
  signature of ``statistic``. If an ``axis`` parameter is present in
  ``statistic`` but should not be relied on for vectorized calls, users must
  pass option ``vectorized==False`` explicitly.
- `scipy.stats.multivariate_normal` will now raise a ``ValueError`` when the
  covariance matrix is not positive semidefinite, regardless of which method
  is called.

*******
Authors
*******

> * h-vetinari (10)

Reviewer 2: I know it's not new, but maybe mention (here and in the template) what the number in parentheses is. Commits, I'm guessing?
